lib/ws_deque.ml

```diff
@@ -0,0 +1,196 @@
+(*
+ * Copyright (c) 2015, Théo Laurent <theo.laurent@ens.fr>
+ * Copyright (c) 2015, KC Sivaramakrishnan <sk826@cl.cam.ac.uk>
+ * Copyright (c) 2017, Nicolas ASSOUAD <nicolas.assouad@ens.fr>
+ * Copyright (c) 2021, Tom Kelly <ctk21@cl.cam.ac.uk>
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+(* Work Stealing Queue
+ *
+ * See:
+ *   Dynamic circular work-stealing deque
+ *   https://dl.acm.org/doi/10.1145/1073970.1073974
+ *  &
+ *   Correct and efficient work-stealing for weak memory models
+ *   https://dl.acm.org/doi/abs/10.1145/2442516.2442524
+ *)
+
+module type S = sig
+  type 'a t
+  val create : unit -> 'a t
+  val is_empty : 'a t -> bool
+  val size : 'a t -> int
+  val push : 'a t -> 'a -> unit
+  val pop : 'a t -> 'a
+  val steal : 'a t -> 'a
+end
+
+module CArray = struct
+
+  type 'a t = {
+    arr  : 'a array;
+    mask : int
+    }
+
+  let rec log2 n =
+    if n <= 1 then 0 else 1 + (log2 (n asr 1))
+
+  let create n v =
+    let sz = Int.shift_left 1 (log2 n) in
+    assert ((sz >= n) && (sz > 0));
+    assert (Int.logand sz (sz-1) == 0);
+    {
+      arr  = Array.init sz (fun _ -> v);
+      mask = sz - 1
+    }
+
+  let size t = Array.length t.arr [@@inline]
+
+  let get t i =
+    Array.unsafe_get t.arr (Int.logand i t.mask) [@@inline]
+
+  let put t i v =
+    Array.unsafe_set t.arr (Int.logand i t.mask) v [@@inline]
+
+  let grow t top bottom =
+    let s = size t in
+    let ns = 2 * s in
+    let out = create ns (Obj.magic ()) in
+    for i = top to bottom do
+      put out i (get t i)
+    done;
+    out
+
+  let shrink t top bottom =
+    let s = size t in
+    let ns = s / 2 in
+    let out = create ns (Obj.magic ()) in
+    for i = top to bottom do
+      put out i (get t i)
+    done;
+    out
+
+end
+
+module M : S = struct
+  let min_size = 32
+  let shrink_const = 3
+
+  type 'a t = {
+    top : int Atomic.t;
+    bottom : int Atomic.t;
+    tab : 'a ref CArray.t Atomic.t;
```

Comment:
From the original Chase-Lev Paper

> Also, unlike all previous algorithms, the top and bottom fields are used merely to indicate the two ends of the deque— no tag field is required to eliminate the ABA problem. 

We're never CASing on the actual values. Hence, IIUC the ref cell is not necessary?

Reply:
I see good performance improvement with the removal of `ref`. 

```bash
(* old *)
$ OCAMLRUNPARAM="v=0x400" hyperfine './_build/default/test/task_throughput.exe 24 32768 1_000_000'
Benchmark #1: ./_build/default/test/task_throughput.exe 24 32768 1_000_000
  Time (mean _ _):      5.253 s _  0.111 s    [User: 125.769 s, System: 0.089 s]
  Range (min _ max):    4.977 s _  5.364 s    10 runs

(* new *)
 OCAMLRUNPARAM="v=0x400" hyperfine './_build/default/test/task_throughput.exe 24 32768 1_000_000'
Benchmark #1: ./_build/default/test/task_throughput.exe 24 32768 1_000_000
  Time (mean _ _):      4.469 s _  0.149 s    [User: 106.980 s, System: 0.039 s]
  Range (min _ max):    4.089 s _  4.600 s    10 runs
```
 
This is on an untuned IITM machine with 28 cores.

Reply:
The `ref` is there to ensure that we can break the link for GC. 

If we don't have a `ref`, then all the values in the backing array are reachable for the GC. Because ownership of the values in the array is handled via `top` and `bottom`, you can not write to the backing array after a `steal` to break the GC link. That is the reason for the inclusion of the `ref`.
 
See: https://github.com/ocaml-multicore/domainslib/pull/29#issuecomment-849713074

Reply:
This makes sense to me. Clever trick. 👍